### PR TITLE
[SPARK-45401][PYTHON] Add a new method `cleanup` in the UDTF interface

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR adds a new API `cleanup` to the current Python UDTF interface.

### Why are the changes needed?

Currently, the `terminate` method of a UDTF is always executed, regardless of whether the `eval` method calls are successful. This is problematic as users might have certain logic in `terminate` that can only be executed once all `eval` calls succeed.

But what if users wish to perform cleanup actions during UDTF execution, such as closing connections? One option is for users to embed a try...except logic within the eval call:

```
def eval(self, row: Any):
  try:
    run_code()
  except Exception:
    clean_up()
```

However, running this try-except block for every eval call can be expensive, potentially affecting the performance of UDTFs.

To tackle this, we can introduce a new method in the UDTF interface that will be called regardless of the outcome. Now the logic would look like:

```
try:
  for row in rows:
    eval(row)
  terminate()
finally:
  cleanup()
```

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Unit test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @ueshin @dtenedor

File: python/pyspark/sql/tests/test_udtf.py [hunk 1]
```diff
@@ -406,6 +406,35 @@ def terminate(self):
...
+        with tempfile.TemporaryDirectory() as d:
+            path = os.path.join(d, "file.txt")
+
+            @udtf(returnType="x: int")
+            class TestUDTF:
+                def __init__(self):
+                    self.path = path
+
+                def eval(self, x: int):
+                    raise Exception("eval error")
```
Comment (dtenedor) [inline python/pyspark/sql/tests/test_udtf.py:419]: can you add another test case with an exception from `terminate` as well? Then `cleanup` should still get called?

**Review by dtenedor — APPROVED** [review]
Looks great, thanks for implementing this!

**Review by ueshin — APPROVED** [review]

**Comment by allisonwang-db** [comment]
I think we should also merge this to spark 3.5

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @alismess-db and all.
Merged to master.

**Pull request closed without merging**